core/trino-main/src/main/java/io/trino/sql/planner/QueryPlanner.java

```diff
@@ -1093,6 +1106,76 @@ private FrameBoundPlanAndSymbols planFrameBound(PlanBuilder subPlan, PlanAndMapp
         return new FrameBoundPlanAndSymbols(subPlan, Optional.of(frameBoundSymbol), sortKeyCoercedForFrameBoundComparison);
     }
 
+    private FrameOffsetPlanAndSymbol planFrameOffset(PlanBuilder subPlan, Optional<Symbol> frameOffset)
+    {
+        if (frameOffset.isEmpty()) {
+            return new FrameOffsetPlanAndSymbol(subPlan, Optional.empty());
+        }
+
+        Symbol offsetSymbol = frameOffset.get();
+        Type offsetType = symbolAllocator.getTypes().get(offsetSymbol);
+
+        // Append filter to validate offset values. They mustn't be negative or null.
+        Expression zeroOffset = zeroOfType(offsetType);
+        ResolvedFunction fail = metadata.resolveFunction(QualifiedName.of("fail"), fromTypes(VARCHAR));
+        Expression predicate = new IfExpression(
+                new ComparisonExpression(GREATER_THAN_OR_EQUAL, offsetSymbol.toSymbolReference(), zeroOffset),
+                TRUE_LITERAL,
+                new Cast(
+                        new FunctionCall(
+                                fail.toQualifiedName(),
+                                ImmutableList.of(new Cast(new StringLiteral("Window frame offset value must not be negative or null"), toSqlType(VARCHAR)))),
+                        toSqlType(BOOLEAN)));
+        subPlan = subPlan.withNewRoot(new FilterNode(
+                idAllocator.getNextId(),
+                subPlan.getRoot(),
+                predicate));
+
+        if (offsetType.equals(BIGINT)) {
+            return new FrameOffsetPlanAndSymbol(subPlan, Optional.of(offsetSymbol));
+        }
+
+        Expression offsetToBigint;
+
+        if (offsetType instanceof DecimalType && !((DecimalType) offsetType).isShort()) {
+            String maxBigint = Long.toString(Long.MAX_VALUE);
+            int maxBigintPrecision = maxBigint.length();
+            int actualPrecision = ((DecimalType) offsetType).getPrecision();
+
+            if (actualPrecision < maxBigintPrecision) {
+                offsetToBigint = new Cast(offsetSymbol.toSymbolReference(), toSqlType(BIGINT));
+            }
+            else if (actualPrecision > maxBigintPrecision) {
+                // In case of long overflow, truncate to max bigint
+                offsetToBigint = new GenericLiteral("BIGINT", maxBigint);
```

Comment:
Why truncate? This could theoretically result in incorrect results, no? If so, it'd be better to just fail with a "out of bounds" error.

Reply:
If the offset value exceeds max bigint, it implies that the frame bound falls beyond partition bound. In such case, there is no error, but the frame bound is set to the partition bound. Passing max bigint as the offset value has the same effect.

Reply:
ok, please add a comment with that explanation